tests/conftest.py

```diff
@@ -55,9 +56,118 @@ def pytest_addoption(parser):
     )
 
 
+def _modify_compose_for_testing(request):
+    demo = "../docker-compose.demo.yml"
+    testing = "../docker-compose.testing.yml"
+    storage_proxy_demo = "../storage-proxy/docker-compose.storage-proxy.demo.yml"
+    storage_proxy_testing = "../storage-proxy/docker-compose.storage-proxy.testing.yml"
+
+    # Published ports in docker-compose-demo and docker compose.storage-proxy.demo
+    demo_ports = ["9000:9000", "8080:8080", "443:443", "80:80"]
+    demo_storage_proxy_ports = ["9000:9000"]
+
+    # Remove published ports
+    for ports, out, file in zip(
+        [demo_ports, demo_storage_proxy_ports],
+        [testing, storage_proxy_testing],
+        [demo, storage_proxy_demo],
+    ):
+
+        command = ["sed"]
+        for port in ports:
+            command.extend(["-e", f"/{port}/d"])
+        subprocess.run(command + ["-e", "/ports:/d", file], stdout=open(out, "w"))
+
+    # Disable download speed limits
+    subprocess.run(["sed", "-e", "s/DOWNLOAD_SPEED/#DOWNLOAD_SPEED/", "-i", testing])
+
+    # Whitelist all IPs/DNS names in the gateway
+    subprocess.run(
+        ["sed", "-e", "s/ALLOWED_HOSTS: .*/ALLOWED_HOSTS: ~./", "-i", testing]
+    )
+
+
+@pytest.fixture(scope="session", autouse=True)
+def modify_compose_for_testing(request):
+    return _modify_compose_for_testing(request)
+
+
+def _extract_fs_from_image(request, client_compose_file):
+    d = tempfile.TemporaryDirectory()
+
+    def cleanup():
+        shutil.rmtree(d.name, ignore_errors=True)
+
+    request.addfinalizer(cleanup)
+
+    image_type = "mender-client"
+    if "gateway" in client_compose_file:
+        image_type = "mender-gateway"
+
+    with open(client_compose_file) as f:
+        data = yaml.safe_load(f)
+        image = data["services"][image_type]["image"]
```

Comment:
Observation: It works here, since we _know_ (at least for know) - that services will always be present. But in general we prefer to be as defensive as possible. Can you think of a way to guard this, and then fail graciously if any of your assumptions here are wrong? :)